### ISSUE TITLE
Simulator of disaggregated domestic electricity demand at 1 Hz

### ISSUE THREAD

**Issue by JackKelly** [nilmtk/nilmtk#26]
We're not planning to build the simulator for a while but I wanted to start this feature request so we can keep notes about the simulator.

Some specific feature ideas:
- Test the effect of:
  - change sample rate
  - multiple appliance on at same time
  - different types of appliance, etc
- Generating scenarios:
  - takes a simple text script as its input (purely deterministic) (Text scripts would allow researchers to share their “scenarios”.)
  - nice GUI for creating these text scripts
  - sample randomly from the probability distributions of time-of-use of each appliance that we calculate from real datasets, so it can automatically create an endless stream of realistic simulations
#### References

If we want to go down the route of trying to simulate appliances themselves (rather than stitching together appliance waveforms) then Dr Mark Bilton's PhD thesis might be of interest... he built a simulator for appliances down to a remarkable level of detail.  Appliances are modelled pretty much from first principals.  Appliances are described in terms of a FSM, volumes of physical containers (to simulate heating and cooling), U-values of insulators for fridges, etc etc.  Very very detailed stuff. http://www.mendeley.com/c/6313531404/g/3295941/bilton-2011-electricity-demand-measurement-modelling-and-management-of-uk-homes/ 

And here's the Barker et al paper that Nipun mentioned: "Empirical Characterization and Modeling of Electrical Loads in Smart Homes". http://lass.cs.umass.edu/papers/pdf/igcc13-model.pdf

**Comment by JackKelly** [comment]
Has anyone come across a domestic electricity simulator which can simulate power demand at one-second intervals?

If not then I'm toying with the idea of building something.  It might work something like this:

First, build simple statistical models of when appliances are used.  These models would be trained from long-duration but low resolution datasets like HES and Dataport.  The models would try to capture things like the time of day that appliances are used (e.g. toasters are often used in the morning); the correlation between appliances (e.g. toasters and kettles are often used within a few minutes of each other), the correlation between external temperature and appliances, etc.

You'd build different statistical models for each country.  And, for each country, it would probably be necessary to create several models, for several different usage patterns (e.g. homes with young children; single-occupancy homes where the occupant is away during the week days etc.)

Then the system would extract the 1Hz power demand from real appliances from datasets such as Tracebase.

To simulate power data at 1Hz, the system would then sample randomly from the statistical model of appliance usage, conditional on real historical temperature data (from a time period which was not in the original data).  Each artificial "house" would be made up of a collection of appliances, taken from the high frequency datasets.

To get even more sophisticated, you could imagine using something like Energy+ to model the building physics in order to model the internal temperature of the home.

Any thoughts?

**Comment by oliparson** [comment]
I remember reading a few papers that either focus on simulating households or evaluate their algorithms using simulated data. It might be worth carrying out a quick literature search first to check if there are any ideas you want to build on, e.g. http://dl.acm.org/citation.cfm?id=2559630

I doubt you'll find open source code for the whole solution though.

Also, I think NILM-Metadata project should come in handy here :)

**Comment by JackKelly** [comment]
> It might be worth carrying out a quick literature search first

yeah.  I did a quick text search for "simulat" in my ref database and found 74 results :)  (not all relevant)

**Comment by marioberges** [comment]
How about Gridlab-D?
http://www.gridlabd.org/

**Comment by JackKelly** [comment]
Thanks Mario, I'll look into Gridlab-D in more detail.  It does say it can do sub-second resolution, which is promising.  I wonder if it provides household-level power demand, though?  And how realistic the appliance power signatures are.  I'm just thinking out loud... I'll look into these questions myself if I do set out to build a simulator...

**Comment by JackKelly** [comment]
Also see the [CBECC-Res Compliance Software Project](http://www.bwilcox.com/BEES/BEES.html)

**Comment by JackKelly** [comment]
A few other ideas for a simulator:

Two separate "modules":
- One for generating "appliance activity scripts".  These "scripts" can then be used to run real automated appliances (e.g. the labs set up by EPRI and PNNL).  Or these same scripts could be used to drive computer-simulated power demand time series.
- A second module which takes "appliance activity scripts" and produces computer-simulated power demand time series.

Could imagine setting up a web service where people just download freshly simulated data, with whatever parameters they require.

**Comment by JackKelly** [comment]
Would also be useful to be able to simulate "abnormal" (but physically plausible) behaviour.  The sort of behaviour which *does* occur in real homes and which would provide an excellent opportunity to reduce energy consumption via feedback.  e.g.:

* a fridge with a broken door seal
* stuff which is accidentally left on 24/7 and shouldn't be: 
  - electric immersion heaters
  - electric swimming pool heater
  - electric food oven
  - a DVR which never switches off
* an ancient fridge

**Comment by JackKelly** [comment]
A few other quick notes on ideas which have come up during the NILM Workshop:

* The simulator needs to be especially accurate at peak demand times.  One use case for NILM is figuring out which appliances are contributing to peak demand.
* We should put effort into metrics for measuring how "realistic" the simulated data is. Not entirely clear how to compute this!

**Comment by JackKelly** [comment]
I've started sketching out some very rough design ideas for the simulator on [this Google doc](https://docs.google.com/document/d/1xv7VcQWfVk32BBVOU5gL7LFuqcrSaO_OfLWRYs36slA/edit?usp=sharing).  Comments welcome.

**Comment by oliparson** [comment]
https://people.umass.edu/~dongchen/smartgridcomm16.pdf

**Comment by JackKelly** [comment]
Thanks Oli, I've added Dong Chen's paper to [the list of references on the Google doc](https://docs.google.com/document/d/1xv7VcQWfVk32BBVOU5gL7LFuqcrSaO_OfLWRYs36slA/edit#heading=h.cpfdo5q2bmkj).

**Comment by franchesoni** [comment]
Awesome! With my research group we are getting into NILM and probably implement some simulations. What's the state-of-the-art? Last comment is from 2016... Is there an implementation?

**Comment by PMeira** [comment]
@franchesoni I plan to work on this after NILMTK v0.3 is released, which is delayed a bit due to the delayed Pandas v0.24 release. 

As you can see in this thread, here are multiple third-party simulator that could be used to generate that but I'm not aware is straightforward. I did some work on the topic but nothing really useful yet.
